---
title: "single cell cluster ananse analysis"
output:
  html_document:
---

```{r setup}
# if (!requireNamespace("remotes", quietly = TRUE)) {
#   install.packages("remotes")
# }
# Sys.unsetenv("GITHUB_PAT")
# remotes::install_github("mojaveazure/seurat-disk", upgrade = "never")#update none
# remotes::install_github("JGASmits/AnanseSeuratWrapper")
# install.packages("BiocManager")
# BiocManager::install("ComplexHeatmap", update = F)
# BiocManager::install("dendextend", update = F)


library(Signac)
library(Seurat)
library(EnsDb.Hsapiens.v86)
library(BSgenome.Hsapiens.UCSC.hg38)
library(SeuratDisk)
library(stringr)
#library(AnanseSeuratWrapper)
library(ComplexHeatmap)
library(circlize)
set.seed(1234)
knitr::opts_knit$set(root.dir = normalizePath(".."))
```

```{r}
# load the RNA and ATAC data
#counts <- Read10X_h5('/scratch/jsmits/scANANSE/data/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5')
counts <- Read10X_h5('./scANANSE/data/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5')
fragpath <- "./scANANSE/data/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz"
# get gene annotations for hg38
annotation <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)
seqlevelsStyle(annotation) <- "UCSC"
# create a Seurat object containing the RNA adata
pbmc <- CreateSeuratObject(
  counts = counts$`Gene Expression`,
  assay = "RNA"
)
# create ATAC assay and add it to the object
pbmc[["ATAC"]] <- CreateChromatinAssay(
  counts = counts$Peaks,
  sep = c(":", "-"),
  fragments = fragpath,
  annotation = annotation
)

DefaultAssay(pbmc) <- "ATAC"
pbmc <- NucleosomeSignal(pbmc)
pbmc <- TSSEnrichment(pbmc)

# filter out low quality cells
pbmc <- subset(
  x = pbmc,
  subset = nCount_ATAC < 100000 &
    nCount_RNA < 25000 &
    nCount_ATAC > 1000 &
    nCount_RNA > 1000 &
    nucleosome_signal < 2 &
    TSS.enrichment > 1
)

# call peaks using MACS2
peaks <- CallPeaks(pbmc)
# remove peaks on nonstandard chromosomes and in genomic blacklist regions
peaks <- keepStandardChromosomes(peaks, pruning.mode = "coarse")
peaks <- subsetByOverlaps(x = peaks, ranges = blacklist_hg38_unified, invert = TRUE)

# quantify counts in each peak
macs2_counts <- FeatureMatrix(
  fragments = Fragments(pbmc),
  features = peaks,
  cells = colnames(pbmc)
)

# create a new assay using the MACS2 peak set and add it to the Seurat object
pbmc[["peaks"]] <- CreateChromatinAssay(
  counts = macs2_counts,
  fragments = fragpath,
  annotation = annotation
)

DefaultAssay(pbmc) <- "RNA"
pbmc <- SCTransform(pbmc)
pbmc <- RunPCA(pbmc)

DefaultAssay(pbmc) <- "peaks"
pbmc <- FindTopFeatures(pbmc, min.cutoff = 5)
pbmc <- RunTFIDF(pbmc)
pbmc <- RunSVD(pbmc)
```

```{r}
# load PBMC reference
reference <- LoadH5Seurat("./scANANSE/data/pbmc_multimodal.h5seurat")
DefaultAssay(pbmc) <- "SCT"

# transfer cell type labels from reference to query
transfer_anchors <- FindTransferAnchors(
  reference = reference,
  query = pbmc,
  normalization.method = "SCT",
  reference.reduction = "spca",
  recompute.residuals = FALSE,
  dims = 1:50
)

predictions <- TransferData(
  anchorset = transfer_anchors, 
  refdata = reference$celltype.l2,
  weight.reduction = pbmc[['pca']],
  dims = 1:50
)

pbmc <- AddMetaData(
  object = pbmc,
  metadata = predictions
)

# set the cell identities to the cell type predictions
Idents(pbmc) <- "predicted.id"

# set a reasonable order for cell types to be displayed when plotting
levels(pbmc) <- c("CD4 Naive", "CD4 TCM", "CD4 CTL", "CD4 TEM", "CD4 Proliferating",
                  "CD8 Naive", "dnT",
                 "CD8 TEM", "CD8 TCM", "CD8 Proliferating", "MAIT", "NK", "NK_CD56bright",
                 "NK Proliferating", "gdT",
                 "Treg", "B naive", "B intermediate", "B memory", "Plasmablast",
                 "CD14 Mono", "CD16 Mono",
                 "cDC1", "cDC2", "pDC", "HSPC", "Eryth", "ASDC", "ILC", "Platelet")
```
```{r}
# build a joint neighbor graph using both assays
pbmc <- FindMultiModalNeighbors(
  object = pbmc,
  reduction.list = list("pca", "lsi"), 
  dims.list = list(1:50, 2:40),
  modality.weight.name = "RNA.weight",
  verbose = TRUE
)

# build a joint UMAP visualization
pbmc <- RunUMAP(
  object = pbmc,
  nn.name = "weighted.nn",
  assay = "RNA",
  verbose = TRUE
)
DimPlot(pbmc, label = TRUE, repel = TRUE, reduction = "umap") + NoLegend()
```
```{r}
Idents(pbmc) <- str_replace(Idents(pbmc), ' ', '-')
Idents(pbmc) <- str_replace(Idents(pbmc), '_', '-')
pbmc$predicted.id <- str_replace(pbmc$predicted.id, ' ', '-')
pbmc$predicted.id <- str_replace(pbmc$predicted.id, '_', '-')
```


```{r}
rds_file <- './scANANSE/preprocessed_PDMC.Rds'
if(file.exists(rds_file)){
  pdmc <- readRDS(rds_file)
}else{saveRDS(pbmc, file = rds_file)}

pdf('./scANANSE/umap.pdf',width=7,height=3.5,paper='special') 
DimPlot(pbmc, label = TRUE, repel = TRUE, reduction = "umap") + NoLegend()
dev.off()

table(pbmc@meta.data$predicted.id)
```


```{r}
#source('/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/markdown_notebooks/R_functions/ANANSE_seurat_wrapper_V2.R')
#Load pre-processed seurat object RDS file
rds_file <- './scANANSE/preprocessed_PDMC.Rds'
pbmc <- readRDS(rds_file)

export_CPM_scANANSE(pbmc,
                    min_cells <- 25,
                    output_dir ='./scANANSE/analysis',
                    cluster_id = 'predicted.id',
                    RNA_count_assay = 'RNA')

export_ATAC_scANANSE(pbmc,
                     min_cells <- 25,
                     output_dir ='./scANANSE/analysis',
                    cluster_id = 'predicted.id',
                    ATAC_peak_assay= 'peaks')

#specify additional contrasts: 
contrasts <-  list('B-naive_B-memory',
                   'B-memory_B-naive',
                   'B-naive_CD16-Mono',
                   'CD16-Mono_B-naive')

config_scANANSE(pbmc,
                min_cells <- 25,
                output_dir ='./scANANSE/analysis',
                cluster_id = 'predicted.id',
                genome = './scANANSE/data/hg38',
                additional_contrasts = contrasts)

DEGS_scANANSE(pbmc,
              min_cells <- 25,
              output_dir ='./scANANSE/analysis',
              cluster_id = 'predicted.id',
              additional_contrasts = contrasts)

```

## Run Anansnake before processing
Lets load the results from Anansnake
```{r}
source("/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/AnanseSeurat/R/ANANSE_seurat_wrapper.R")

rds_file <- './scANANSE/preprocessed_PDMC.Rds'
pbmc <- readRDS(rds_file)
cluster_id = 'predicted.id'

setwd("/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/")
pbmc <- import_seurat_scANANSE(pbmc,
                               cluster_id = 'predicted.id',
                               anansnake_inf_dir= "./scANANSE/analysis2/influence/")

TF_influence <- per_cluster_df(pbmc, 
                               cluster_id = 'predicted.id', 
                               assay = 'influence')

```

get top 5 highest TFs
```{r}
TF_influence$TF <- rownames(TF_influence)
TF_long <- reshape2::melt(TF_influence, id.vars = 'TF')
TF_influence$TF <- NULL
colnames(TF_long) <- c('TF','cluster', 'influence')
TF_long <- TF_long[order(TF_long$influence, decreasing = TRUE), ]

#get the top n TFs per cluster
topTF <- Reduce(rbind,                                 
                    by(TF_long,
                       TF_long["cluster"],
                       head,
                       n = 5))# Top N highest TFs by cluster

top_TFs <- unique(topTF$TF)
  
TF_table <- topTF %>% 
  dplyr::group_by(cluster) %>% 
  dplyr::mutate('TopTFs' = paste0(TF, collapse = " ")) 

unique(TF_table[,c('cluster','TopTFs')])
```

```{r}
library(ComplexHeatmap)
library(circlize)
col_fun = colorRamp2(c(0, 1), c("white", "orange"))
mat <- TF_influence[rownames(TF_influence) %in% top_TFs,]

#pdf('./scANANSE/analysis/ANANSE_Heatmap.pdf',width=16,height=8,paper='special')
Heatmap(mat, col = col_fun)
#dev.off()
```


Generate a Heatmap of the top TFs
```{r}
set.seed(123)
DefaultAssay(object = pbmc) <- "influence"
mat <- TF_influence[rownames(TF_influence) %in% top_TFs,]

col_fun = colorRamp2(c(0, 1), c("white", "orange"))
costum_sample_order = c('CD14-Mono','CD16-Mono','cDC2','pDC','HSPC','B-naive','B-intermediate','B-memory','CD8-Naive','CD4-Naive','CD8-TCM','CD4-TEM','CD4-TCM','Treg','MAIT','CD8-TEM','gdT','NK')

dend <- stats::as.dendrogram(stats::hclust(dist(t(mat))))
#dend <-dendextend::rotate(dend,costum_sample_order) #rotate to match labels new order}

#pdf("./scANANSE/analysis/ANANSE_Heatmap.pdf" ,width=16,height=8,paper='special')
print(ComplexHeatmap::Heatmap(mat,
                        row_dend_side = "right",
                        show_column_dend = T,
                        col = col_fun,
                        cluster_columns = dend,
                        row_km_repeats = 100))
#dev.off()
```

```{r}
pdf('./scANANSE/analysis/reference_umap.pdf',width=15,height=15,paper='special') 
Idents(object = reference) <- "celltype.l2"
print(DimPlot(reference, label = T, repel = TRUE, reduction = "umap") )
Idents(object = reference) <- "celltype.l1"
print(DimPlot(reference, label = T, repel = TRUE, reduction = "umap") )
dev.off()

pdf('./scANANSE/analysis/ANANSE_umap.pdf',width=10,height=5,paper='special') 
Idents(object = pbmc) <- "predicted.id"
print(DimPlot(pbmc, label = T, repel = TRUE, reduction = "umap")+ NoLegend())
dev.off()
```


```{r}
highlight_TF1 <- c('STAT4','LEF1','MEF2C')

#pdf('./scANANSE/analysis/ANANSE_highlight.pdf',width=10,height=10,paper='special') 
DefaultAssay(object = pbmc) <- "RNA"
plot_expression1 <- FeaturePlot(pbmc, features = highlight_TF1, ncol = 1)
DefaultAssay(object = pbmc) <- "influence"
plot_ANANSE1 <- FeaturePlot(pbmc,ncol = 1, features = highlight_TF1, cols = c("darkgrey", "#fc8d59"))
print(DimPlot(pbmc, label = T, repel = TRUE, reduction = "umap")+ NoLegend())
print(plot_expression1|plot_ANANSE1)
#dev.off()
```


```{r}
# TF_df_celltype <- as.data.frame(t(TF_influence_scores))
# TF_df_celltype[['cell_type']] <- rownames(TF_df_celltype)
# 
# pdf('./scANANSE/analysis/ANANSE.pdf' ,width=15,height=15,paper='special') 
# plot_clustering <- DimPlot(pbmc, reduction = "umap", label = TRUE) + NoLegend()
# for (TF in top_TFs){
# print(TF)
# DefaultAssay(object = pbmc) <- "RNA"
# plot_expression <- FeaturePlot(pbmc, features = TF)+ ggplot2::ggtitle(paste0(TF,' expression'))
# DefaultAssay(object = pbmc) <- "ANANSE"
# plot_ANANSE <- FeaturePlot(pbmc, features = TF, cols = c("grey", "purple")) + ggplot2::ggtitle(paste0(TF,' ANANSE influence'))
# plot_ANANSE_bar <- ggplot2::ggplot(TF_df_celltype, ggplot2::aes(x=cell_type, y= .data[[TF]])) + ggplot2::geom_bar(stat='identity') + ggplot2::ylab(paste0(TF, ' influence score')) + ggplot2::coord_flip()
# print(plot_clustering/plot_ANANSE_bar|plot_expression/plot_ANANSE)
# 
# #print(plot_clustering/plot_ANANSE_bar|plot_expression/plot_ANANSE)
# }
# dev.off()
```

Lets finally also output the ATAC counts for motif enrichment
```{r}
export_ATAC_maelstrom(pbmc,
              min_cells <- 25,
              output_dir ='./scANANSE/analysis',
              cluster_id = 'predicted.id')
```

#run gimme maelstrom using:
```{bash}
#conda create -n gimme python=3 gimmemotifs
# Activate the environment
#conda activate gimme
#gimme maelstrom scANANSE/analysis/Peaks_scaled.tsv hg38 scANANSE/analysis/maelstrom

```

import motif enrichment:
```{r}
rds_file <- './scANANSE/preprocessed_PDMC.Rds'
pbmc <- readRDS(rds_file)
#is.null(pbmc@assays[['test']])

source("/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/AnanseSeurat/R/ANANSE_seurat_wrapper.R")
pbmc <- import_seurat_maelstrom(pbmc,
              cluster_id = 'predicted.id',
              maelstrom_dir = './scANANSE/analysis/maelstrom/')

motif_scores <- per_cluster_df(pbmc,
                                 assay = 'maelstrom',
                                 cluster_id = 'predicted.id')

```

```{r}
source("/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/AnanseSeurat/R/ANANSE_seurat_wrapper.R")
test <- Maelstrom_Motif2TF(pbmc,
                         cluster_id = 'predicted.id',
                         maelstrom_dir = './scANANSE/analysis/maelstrom/',
                         RNA_expression_assay = "SCT",
                         output_dir ='./scANANSE/analysis',
                         expr_tresh = 10,
                         cor_tresh = 0.3,
                         combine_motifs = 'mean',
                         )
```


```{r}
 pdf('./scANANSE/analysis/repressive_motifs.pdf' ,width=15,height=45,paper='special') 
#highlight_TF1 <- c('PAX5','STAT6','MEF2A','TFCP2','PRDM1')
highlight_TF1 <- var_repr
DefaultAssay(object = pbmc) <- "RNA"
plot_expression1 <- FeaturePlot(pbmc, features = highlight_TF1, ncol = 1)
DefaultAssay(object = pbmc) <- "MotifTFcor"
plot_Maelstrom <- FeaturePlot(pbmc,ncol = 1, features = highlight_TF1, cols = c("darkred","darkgrey"))
print(plot_expression1|plot_Maelstrom)
dev.off()
```



```{r}
source("/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/AnanseSeurat/R/ANANSE_seurat_wrapper.R")
m2f_df_match <- generate_TF_Motif_tables(list(pbmc,t(motif_scores)),
                  cluster_id = 'predicted.id',
                  maelstrom_dir = './scANANSE/analysis/maelstrom/',
                  output_dir ='./scANANSE/analysis',
                  expr_tresh = 100,
                  cor_tresh = 0.3,
                  m2f_df = NULL,
                  repressors = T
                  )


View(m2f_df_match[['scaled_motif_score']])


repressive_score <- per_cluster_df(pbmc,
                                 assay = 'repressorScore',
                                 cluster_id = 'predicted.id')

activitor_score <- per_cluster_df(pbmc,
                                 assay = 'activatorScore',
                                 cluster_id = 'predicted.id')

```


```{r}
highlight_TF1 <- c('PAX5','CEBPG')
DefaultAssay(object = pbmc) <- "RNA"
plot_expression1 <- FeaturePlot(pbmc, features = highlight_TF1, ncol = 1)
DefaultAssay(object = pbmc) <- "repressorScore"
plot_Maelstrom <- FeaturePlot(pbmc,ncol = 1, features = highlight_TF1, cols = c("darkred", "darkgrey"))
print(plot_expression1|plot_Maelstrom)
```
```{r}
generate_TF_Motif_tables_reb <- function(maelstrom_object,
                                     m2f_df = NULL, 
                                     cluster_id = 'seurat_clusters',
                                     maelstrom_dir = '/ceph/rimlsfnwi/data/moldevbio/veenstra/snabel/scrna/CMdiff_10XMO_EB_d8/R_analysis/20221107_TFMotifVisualization/runmaelstrom/',
                                     RNA_expression_assay = "RNA",
                                     RNA_expression_slot = "data",
                                     repressors = FALSE,
                                     expr_tresh = 10,
                                     cor_tresh = 0.30,
                                     cor_method = "pearson",
                                     output_dir = "./",
                                     select_most_var = FALSE){
  ## Check if m2f_df object provided or path to Maelstrom output.
  ## Check if m2f_df object provided contains the right columns.
  
  m2f_df_cols <- c("Motif","Factor")
  if (is.null(m2f_df)){
    if (is.null(maelstrom_dir)){
      print("Provide path for maelstrom_dir or provide table m2f_df with Motif and Factor column.")
    } else {m2f_file <- paste0(maelstrom_dir, "/nonredundant.motifs.motif2factors.txt")
    m2f_df <- utils::read.table(m2f_file, header = T, sep = '\t',check.names=FALSE)
    }
  } else if (!unique(m2f_df_cols %in% colnames(m2f_df))){
    print("Provide m2f_df with at least 2 columns with names Motif and Factor.")
  }
  
  ## Load needed objects
  seurat_object <- maelstrom_object[[1]]
  mot_mat <- t(maelstrom_object[[2]])
  
  ## Set up Seurat object
  Seurat::DefaultAssay(seurat_object) <- RNA_expression_assay
  genes_expressed <- rownames(seurat_object)[rowSums(seurat_object[[RNA_expression_assay]]@counts) >= expr_tresh]
  seurat_object <- seurat_object[genes_expressed,]
  
  ## Select motifs with binding TFs present in object
  m2f_df <- m2f_df[m2f_df$Factor %in% rownames(seurat_object),]
  
  ## Check if data is normalized
  if (identical(seurat_object[[RNA_expression_assay]]@data, seurat_object[[RNA_expression_assay]]@counts) & RNA_expression_slot == "data"){
    print("Your data slot was not yet normalized.")
    print(paste0("Seurat NormalizeData with default settings will be run on all the genes in the ", RNA_expression_assay, " assay."))
    seurat_object <- Seurat::NormalizeData(seurat_object, assay = RNA_expression_assay)
  }
  
  ## Obtain df with mean expression
  exp_mat <- Seurat::AverageExpression(seurat_object, assays = RNA_expression_assay,
                                       slot = RNA_expression_slot, 
                                       features = m2f_df$Factor,
                                       group.by = cluster_id)[[1]]
  
  ## make sure that all genes in matrix have mean expression > 0
  exp_mat <- exp_mat[!rowSums(exp_mat) <= 0,]
  ## Select the same exp_mat columns as in mot_mat columns (if the grouping var is the same)
  exp_mat <- exp_mat[,colnames(mot_mat)]
  
  ## limit table to motifs and TFs present in dataset
  mot_mat <- mot_mat[rownames(mot_mat) %in% m2f_df$Motif,]
  TF_mat <- exp_mat[rownames(exp_mat) %in% m2f_df$Factor,]
  
  m2f_df_match <- m2f_df[m2f_df$Motif %in% rownames(mot_mat) & m2f_df$Factor %in% rownames(TF_mat),]
  
  m2f_df_match$cor <- NA
  ## perform correlations between cluster expression and cluster motif enrichment
  for (i in 1:nrow(m2f_df_match)){
    m2f_df_match$cor[i] <- stats::cor(mot_mat[m2f_df_match$Motif[i],],exp_mat[m2f_df_match$Factor[i],], method = cor_method)
  }
  
  ## Only keep motif-TF combinations with an absolute R higher than treshold
  print(paste0("Only keep motif-TF combinations with an R > ", cor_tresh))
  m2f_df_match <- m2f_df_match[base::abs(m2f_df_match$cor) > cor_tresh,]
  
  # ## Select highest correlating TF per motif
  # m2f_df_unique <- m2f_df_match %>% dplyr::group_by(m2f_df_match$Motif) %>%
  #   dplyr::arrange(desc(cor)) %>% dplyr::filter(dplyr::row_number() == 1)
  
  ## Select highest absolute correlation of TF and motif
  m2f_df_unique <- m2f_df_match %>% dplyr::group_by(m2f_df_match$Motif) %>%
    dplyr::arrange(dplyr::desc(base::abs(m2f_df_match$cor))) %>% dplyr::filter(dplyr::row_number() == 1)
  
  m2f <- m2f_df_unique
  # Select only positive correlations or only negative correlations (repressors)
  if (repressors == TRUE){
    print("Selecting repressors")
    m2f <- m2f_df_unique[m2f_df_unique$cor < 0,]
    assay_name <- "repressor"
  } else {
    print("Selecting activators")
    m2f <- m2f_df_unique[m2f_df_unique$cor > 0,]
    assay_name <- "activator" 
  }
  
  ## Order motifs according to m2f
  mot_plot <- mot_mat[match(m2f$Motif,rownames(mot_mat)),]
  ## Replace motif name by TF name
  rownames(mot_plot) <- m2f$Factor
  
  ## Make motif score per TF (selecting most variable motif per TF or make mean of all motifs associated). 
  if (select_most_var == TRUE){
    print("Most variable binding motif is selected per TF")
    mot_plot <- cbind(mot_plot, matrix(apply(mot_plot, 1, stats::var)))
    colnames(mot_plot)[length(colnames(mot_plot))]<-"var"
    mot_plot <- mot_plot[order(mot_plot[,"var"], decreasing = T),]
    # Remove duplicated occurences of rowname (ordered on variation)
    mot_plot <- mot_plot[!duplicated(rownames(mot_plot)),]
    mot_plot <- mot_plot[,!colnames(mot_plot) %in% c("var")]
  } else {
    print("Take mean motif score of all binding motifs per TF")
    ## Take mean of motifs linked to the same TF
    mot_plot <- stats::aggregate(mot_plot, list(row.names(mot_plot)), mean)
    mot_plot <- as.data.frame(mot_plot, row.names = mot_plot[,1])[,-1]
  }
  
  ## order expression matrix and motif matrix the same way
  exp_plot <- exp_mat[match(rownames(mot_plot),rownames(exp_mat)),]
  
  exp_plot_scale <- t(scale(t(exp_plot)))
  mot_plot_scale <- t(scale(t(mot_plot)))
  
  expression_file <- paste(output_dir, "expression_means_scaled.tsv", sep = '/')
  utils::write.table(exp_plot_scale, expression_file, sep = '\t', quote = F)
  motif_file <- paste(output_dir, "motif_intensities_scaled.tsv", sep = '/')
  utils::write.table(mot_plot_scale, motif_file, sep = '\t', quote = F)
  matrix_list <- list()
  matrix_list[["expression"]] <- exp_plot
  matrix_list[["motif_score"]] <- mot_plot
  matrix_list[["scaled_expression"]] <- exp_plot_scale
  matrix_list[["scaled_motif_score"]] <- mot_plot_scale
  matrix_list[["motif_tf_correlations"]] <- m2f_df_match
  
  ## Create seurat assay with binding factor assay
  new_assay <- as.data.frame(matrix(data = NA, ncol = length(colnames(seurat_object)), nrow = length(rownames(mot_plot))))
  colnames(new_assay) <- colnames(seurat_object)
  rownames(new_assay) <- rownames(mot_plot)
  for (cluster in colnames(mot_plot)){
    cluster_cells <- colnames(seurat_object[,seurat_object@meta.data[[cluster_id]] == cluster])
    for (TF in rownames(new_assay)){
      new_assay[TF,cluster_cells] <- mot_plot[TF,cluster]
    }
  }
  assay_name <- paste0(assay_name, "Score")
  seurat_object[[assay_name]] <- Seurat::CreateAssayObject(new_assay)
  matrix_list[["seurat_object"]] <- seurat_object
  
  return(matrix_list)
}
```

